pkg/proxy/proxy.go

```diff
@@ -223,16 +226,32 @@ func (p *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	defer frontend.Close()
+	ticker := time.NewTicker(websocketPingInterval)
+
+	defer func() {
+		ticker.Stop()
+		frontend.Close()
+	}()
 
 	errc := make(chan error, 2)
 
 	// Can't just use io.Copy here since browsers care about frame headers.
 	go func() { errc <- copyMsgs(frontend, backend) }()
 	go func() { errc <- copyMsgs(backend, frontend) }()
 
-	// Only wait for a single error and let the defers close both connections.
-	<-errc
+	for {
+		select {
+		case <-errc:
+			// Only wait for a single error and let the defers close both connections.
+			return
+		case <-ticker.C:
+			// Send pings to client to prevent load balancers and other middlemen from closing the connection early
+			err := frontend.WriteControl(websocket.PingMessage, []byte{}, time.Now().Add(websocketTimeout))
```

Comment:
I don't think websocket connections allow for concurrent writes.

Reply:
https://godoc.org/github.com/gorilla/websocket#hdr-Concurrency

I think we're going to need a write lock.

Reply:
Thanks for finding this. I pushed another commit with a mutex.